shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java

```diff
@@ -0,0 +1,109 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.embedding.engine.dart;
+
+import android.support.annotation.NonNull;
+import android.util.Log;
+
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import io.flutter.embedding.engine.FlutterJNI;
+import io.flutter.plugin.common.BinaryMessenger;
+
+/**
+ * Message conduit for 2-way communication between Android and Dart.
+ * <p>
+ * WARNING: THIS CLASS IS EXPERIMENTAL. DO NOT SHIP A DEPENDENCY ON THIS CODE.
+ * IF YOU USE IT, WE WILL BREAK YOU.
+ * <p>
+ * See {@link BinaryMessenger}, which sends messages from Android to Dart
+ * See {@link PlatformMessageHandler}, which handles messages to Android from Dart
+ */
+class DartMessenger implements BinaryMessenger, PlatformMessageHandler {
+  private static final String TAG = "DartMessenger";
+
+  private final FlutterJNI flutterJNI;
+  private final Map<String, BinaryMessenger.BinaryMessageHandler> messageHandlers;
+  private final Map<Integer, BinaryMessenger.BinaryReply> mPendingReplies = new HashMap<>();
+  private int mNextReplyId = 1;
+
+  DartMessenger(@NonNull FlutterJNI flutterJNI) {
+    this.flutterJNI = flutterJNI;
+    this.messageHandlers = new HashMap<>();
+  }
+
+  @Override
+  public void setMessageHandler(String channel, BinaryMessenger.BinaryMessageHandler handler) {
+    if (handler == null) {
+      messageHandlers.remove(channel);
+    } else {
+      messageHandlers.put(channel, handler);
+    }
+  }
+
+  @Override
+  public void send(String channel, ByteBuffer message) {
+    send(channel, message, null);
+  }
+
+  @Override
+  public void send(String channel, ByteBuffer message, BinaryMessenger.BinaryReply callback) {
+    int replyId = 0;
+    if (callback != null) {
+      replyId = mNextReplyId++;
+      mPendingReplies.put(replyId, callback);
+    }
+    if (message == null) {
+      flutterJNI.dispatchEmptyPlatformMessage(channel, replyId);
+    } else {
+      flutterJNI.dispatchPlatformMessage(channel, message, message.position(), replyId);
+    }
+  }
+
+  @Override
+  public void handlePlatformMessage(final String channel, byte[] message, final int replyId) {
+    BinaryMessenger.BinaryMessageHandler handler = messageHandlers.get(channel);
+    if (handler != null) {
```

Comment:
subjective: consider using guard clauses and defining `new BinaryMessenger.BinaryReply()` out of line here to avoid so many levels of nesting.

https://testing.googleblog.com/2017/06/code-health-reduce-nesting-reduce.html

Reply:
There's a competing school of thought here that says you should never return early and that all execution paths should be clearly demarked as blocks (i.e., if/else if/else).  If you feel strongly about it, we should chat about those competing ideas.  I see merits on both sides, but tend to find blocked code to lead to fewer bugs.

Reply:
It's ultimately subjective so feel free to keep this as-is. That said I think there could still be a single point of return and reduced the nesting here by defining the `BinaryReply()` implementation out of the try block.

```
private final class PlatformReply implements BinaryMessenger.BinaryReply {
  private final AtomicBoolean done = new AtomicBoolean(false);
 // ... etc
}

// handlePlatformMessage
    BinaryMessenger.BinaryMessageHandler handler = messageHandlers.get(channel);
    if (handler != null) {
      } try {
        final ByteBuffer buffer = (message == null ? null : ByteBuffer.wrap(message));
        handler.onMessage(buffer, new PlatformReply());
     } catch (
```